File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    int key = 0;
    public int kthSmallest(TreeNode root, int k) {
        key = k;
        return helper(root);
    }

    private int helper(TreeNode root) {
        if (root == null) return 0;
        int val = helper(root.left);
        if (key-- == 1) val = root.val;
        if (val == 0) return helper(root.right);
        else return val;
    }

    public int kthSmallest_1(TreeNode root, int k) {
        Stack<TreeNode> st = new Stack<>();
        TreeNode curr = root;
        while(curr != null || !st.isEmpty()){
            while(curr != null) {
                st.push(curr);
                curr = curr.left;
            }
            curr = st.pop();
            if (k == 1) return curr.val;
            k--;
            curr = curr.right;
        }
        return -1;
    }
}

File: Data Structures & Algorithms/delete-node-in-a-bst/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        def minRightSubtree(root):
            curr = root
            while curr.left:
                curr = curr.left
            return curr
        
        def removeNode(root, key):
            if not root: 
                return 
            if key < root.val:
                root.left = removeNode(root.left, key)
            elif key > root.val:
                root.right = removeNode(root.right, key)
            else:
                if not root.left:
                    return root.right
                elif not root.right:
                    return root.left
                
                minNode = minRightSubtree(root.right)
                root.val = minNode.val
                root.right = removeNode(root.right, minNode.val)

            return root

        return removeNode(root, key)
